web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php

```diff
@@ -45,43 +54,54 @@ public static function create(ContainerInterface $container, array $configuratio
       $plugin_definition,
       $container->get('config.factory'),
       $container->get('logger.factory'),
+      $container->get('database'),
       $container->get('http_client_factory'),
+      $container->get('weather_block.get_set_city_user'),
+      $container->get('cache.default'),
+      $container->get('current_user'),
+      $container->get('weather_block.fetch_api_data'),
     );
   }
 
   /**
    * {@inheritdoc}
    */
   public function build(): array {
-    $user = \Drupal::currentUser();
-    $user_id = $user->id();
-
-    $user_config = $this->configuration['users'][$user_id] ?? [];
-    $cities = $user_config['cities'] ?? [];
-    $api_key = $this->configuration['settings']['key'] ?? '';
-
-    if (!is_array($cities)) {
-      $cities = [];
-    }
+    $user_id = $this->accountProxy->id();
+    $cityArray = $this->cityService->getSavedCityForUser($user_id);
+    $city = implode(', ', $cityArray);
+    $api_key = $this->configuration['settings']['key'];
+    $weather_data = [];
 
-    $cache_id = 'weather_block_data_' . $user_id . '_' . md5(serialize($cities));
+    $cache_id = 'weather_block_data_' . md5($city);
 
-    if (!$cache = \Drupal::cache()->get($cache_id)) {
-      $api_data = $this->getDataFromApi($cities, $api_key);
+    if (!$cache = $this->cacheBackend->get($cache_id)) {
+      $api_data = $this->apiData->getDataFromApi($city, $api_key);
 
       if ($api_data === ['no key']) {
         return [];
       }
+      $tags = $this->getCacheTags();
```

Comment:
Pretty sure block-level tags have nothing to do with you setting the cache for API data.
If anything API cache logic should be in getDataFromApi.

And above you save weather data to your table, but here you are using the default cache backend and this doesn't make sense